src/wp-includes/load.php

```diff
@@ -263,6 +263,59 @@ function wp_get_environment_type() {
 	return $current_env;
 }
 
+/**
+ * Retrieves the current development mode.
+ *
+ * The development mode affects how certain parts of the WordPress application behave, which is relevant when
+ * developing for WordPress.
+ *
+ * Valid developer modes are 'core', 'plugin', 'theme', or an empty string to disable developer mode.
+ *
+ * Developer mode is considered separately from `WP_DEBUG` and {@see wp_get_environment_type()}. It does not affect
+ * debugging output, but rather functional nuances in WordPress.
+ *
+ * @since 6.3.0
+ *
+ * @return string The current development mode.
+ */
+function wp_get_development_mode() {
+	static $current_mode = null;
+
+	if ( ! defined( 'WP_RUN_CORE_TESTS' ) && null !== $current_mode ) {
+		return $current_mode;
+	}
+
+	$valid_modes = array(
+		'core',
+		'plugin',
+		'theme',
+		'',
+	);
+
+	$development_mode = WP_DEVELOPMENT_MODE;
+	if ( ! in_array( $development_mode, $valid_modes, true ) ) {
+		$development_mode = '';
+	}
+
+	/**
+	 * Filters the development mode.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $development_mode The development mode. Valid values are 'core', 'plugin', 'theme', or an empty
+	 *                                 string to signify no developer mode.
+	 */
+	$development_mode = apply_filters( 'wp_development_mode', $development_mode );
```

Comment:
Think there is no need of this filter. Problems:
- The output of `wp_get_development_mode()` will be different before and after the filter is used. This inconsistency may cause pretty big problems down the line.
- Unsure of the use cases for this filter? What can be achieved by dynamically switching the development mode while WP is running? This is also inconsistent with using a constant to set the dev. mode.

Reply:
@azaozz Without a filter, how do we make this work in unit tests? I agree with you the filter doesn't provide much value for production, though even there, having a filter will allow the value to be set more dynamically.

Having a filter here is nothing new. Functions like `wp_doing_ajax()` or `wp_doing_cron()` follow a similar pattern for something that as well should never change within a single request.

I think we can continue other alternatives, but having _only_ a constant here is not a good idea, at least because of tests.

Reply:
After https://github.com/WordPress/wordpress-develop/pull/4565/commits/093d74b535232d8dcc88778b8b90f4c02b27093a, it's now realistically infeasible to even use the filter in a plugin or theme, since `wp_get_development_mode()` is potentially called as early as setting default constants.

So we can either use the filter purely for tests, or we have to find another approach, e.g. an environment variable similar to how it is in `wp_get_environment_type()`. Or any other ideas on how to handle that in a way that allows reasonable testing?